Introduce SSID/PSID in favor of alias (#1)
afrind/warp-draft: draft-ietf-moq-transport.md

each active upstream subscription that matches that namespace, it SHOULD send a
SUBSCRIBE to the publisher that sent the ANNOUNCE.

If a relay receives a Publisher-Chosen Subscription ID in SUBSCRIBE_OK, it

> SHOULD assign this ID to downstream subscriptions for the same track. Since
> relays can aggregate tracks from uncoordinated publishers, it is not always

Author: ```suggestion
SHOULD assign this ID to downstream subscriptions for the same track. Since
subscribers can request tracks from uncoordinated publishers through a single
relay session, it is not always
```

possible to reuse the upstream PSID.  If there is no upstream
Publisher-Chosen Subscription ID, or if the upstream PSID is already in use

> downstream for a different track, the relay SHOULD NOT set a 

Author: ```suggestion
downstream for a different track, the relay SHOULD NOT set a
```

  If a server receives a GOAWAY with a non-zero New Session URI Length it MUST
  terminate the session with a Protocol Violation.

## MAX_SUBSCRIPTION_ID {#message-max-subscribe-id}

A publisher sends a MAX_SUBSCRIPTION_ID message to increase the number of
subscriptions a subscriber can create within a session.

The Maximum Subscription ID MUST only increase within a session, and
receipt of a MAX_SUBSCRIPTION_ID message with an equal or smaller Subscription ID
value is a 'Protocol Violation'.

~~~
MAX_SUBSCRIPTION_ID
{
  Type (i) = 0x15,
  Length (i),
  Subscription ID (i),
}
~~~
{: #moq-transport-max-subscribe-id format title="MOQT MAX_SUBSCRIPTION_ID Message"}

> * Subscription ID: The new Maximum Subscription ID for the session. If a Subscription ID

Author: ```suggestion
* Subscription ID: The new Maximum Subscriber-Chosen Subscription ID for
the session. If a Subscriber-Chosen Subscription ID
```

error of 'Too Many Subscribes'.

MAX_SUBSCRIPTION_ID is similar to MAX_STREAMS in ({{?RFC9000, Section 4.6}}),
and similar considerations apply when deciding how often to send MAX_SUBSCRIPTION_ID.
For example, implementations might choose to increase MAX_SUBSCRIPTION_ID as
subscriptions close to keep the number of subscriptions available to subscribers
roughly consistent.

## SUBSCRIBES_BLOCKED {#message-subscribes-blocked}

The SUBSCRIBES_BLOCKED message is sent when a subscriber would like to begin
a new subscription, but cannot because the Subscription ID would exceed the
Maximum Subscription ID value sent by the peer.  The subscriber SHOULD send only
one SUBSCRIBES_BLOCKED for a given Maximum Subscription ID.

A publisher MAY send a MAX_SUBSCRIPTION_ID upon receipt of SUBSCRIBES_BLOCKED,
but it MUST NOT rely on SUBSCRIBES_BLOCKED to trigger sending a
MAX_SUBSCRIPTION_ID, because sending SUBSCRIBES_BLOCKED is not required.

~~~
SUBSCRIBES_BLOCKED
{
  Type (i) = 0x1A,
  Length (i),
  Maximum Subscription ID (i),
}
~~~
{: #moq-transport-subscribes-blocked format title="MOQT SUBSCRIBES_BLOCKED Message"}

> * Maximum Subscription ID: The Maximum Subscription ID for the session on which the subscriber

Author: ```suggestion
* Maximum Subscription ID: The Maximum Subscriber-Chosen Subscription ID for the
session on which the subscriber
```

~~~
{: #moq-transport-subscribe-format title="MOQT SUBSCRIBE Message"}

> * Subscription ID: The subscriber specified identifier (SSID) used to manage

Author: ```suggestion
* Subscription ID: The Subscriber-Chosen Subscription ID (SSID) used to manage
```

~~~
{: #moq-transport-fetch-ok format title="MOQT FETCH_OK Message"}

> * Subscription ID: Fetch Identifier as defined in {{message-fetch}}.

Author: ```suggestion
* Subscription ID: Subscriber-Chosen Subscription Identifier as defined in
  {{message-fetch}}.
```

{
  Type (i) = 0x19,
  Length (i),
  Subscription ID (i),
  Error Code (i),
  Reason Phrase Length (i),
  Reason Phrase (..),
}
~~~
{: #moq-transport-fetch-error format title="MOQT FETCH_ERROR Message"}

> * Subscription ID: Subscription Identifier as defined in {{message-fetch}}.

Author: ```suggestion
* Subscription ID: Subscriber-Chosen Subscription Identifier as defined in
  {{message-fetch}}.
```

FETCH_CANCEL Message {
  Type (i) = 0x17,
  Length (i),
  Subscription ID (i)
}
~~~
{: #moq-transport-fetch-cancel title="MOQT FETCH_CANCEL Message"}

> * Subscription ID: Subscription Identifier as defined in {{message-fetch}}.

Author: ```suggestion
* Subscription ID: Subscriber-Chosen Subscription Identifier as defined in
  {{message-fetch}}.
```

~~~
OBJECT_DATAGRAM {

>   Subscription Identifier (i),

Author: ```suggestion
  Subscription ID (i),
```

~~~
OBJECT_DATAGRAM_STATUS {

>   Subscription Identifier (i),

Author: ```suggestion
  Subscription ID (i),
```

~~~
{: #object-datagram-status-format title="MOQT OBJECT_DATAGRAM_STATUS"}

> * Subscription Identifier: the Subscriber or Publisher chosen Subscription

Author: ```suggestion
* Subscription ID: the Subscriber or Publisher chosen Subscription
```

### Subgroup Header

When a stream begins with `SUBGROUP_HEADER`, all Objects on the stream
belong to the track requested in the subscription identified by `Subscription
Identifier` and the subgroup indicated by 'Group ID' and `Subgroup ID`.

~~~
SUBGROUP_HEADER {

>   Subscription Identifier (i),

Author: ```suggestion
  Subscription ID (i),
```

  Group ID (i),
  Subgroup ID (i),
  Publisher Priority (8),
}
~~~
{: #object-header-format title="MOQT SUBGROUP_HEADER"}

> * Subscription Identifier: the Subscriber or Publisher chosen Subscription

Author: ```suggestion
* Subscription ID: the Subscriber or Publisher chosen Subscription
```

  Identifier indicating the subscription this Subgroup belongs to. If an

>   endpoint receives a subgroup with an unknown Subscriber-Chosen Subscription
>   Identifier, it MUST close the connection with a Protocol Violation.  If it
>   receives a subgroup with an unknown Publisher-Chosen Subscription Identifier,
>   it MAY abandon the stream, or choose to buffer it for a brief period to handle
>   reordering with the control message that establishes the PSID.  The endpoint
>   SHOULD NOT release stream flow control beyond the SUBGROUP_HEADER until the
>   PSID has been established.  TODO: talk about possible deadlocks.

Author: ```suggestion
  endpoint receives a subgroup with aSubscriber-Chosen Subscription
  Identifier that does not correspond to a subscription it initiated, it MUST
  close the connection with a Protocol Violation.  If it receives a subgroup
  with an unknown Publisher-Chosen Subscription Identifier, it MAY
  abandon the stream, or choose to buffer it for a brief period to handle
  reordering with the control message that establishes the PSID.  The endpoint
  SHOULD NOT release stream flow control beyond the SUBGROUP_HEADER until the
  PSID has been established.  TODO: talk about possible deadlocks.
```
